Sources/RayTracer/RayTracer.Library/IIntersectableTrees/OctTrees/OctTree.cs

```diff
@@ -0,0 +1,477 @@
+    {
+        for (int i = 0; i < 8; i++)
+        {
+            octantList[i] = new List<IIntersectable>();
+        }
+    }
+
+    private void BuildChildNodes(List<IIntersectable>[] octantList, BoundingBox[] octant)
+    {
+        for (int i = 0; i < 8; i++)
+        {
+            if (octantList[i].Count != 0)
+            {
+                _childNodes ??= new OctTree[8];
+                _childNodes[i] = CreateNode(octant[i], octantList[i]);
+                _childNodes[i]?.BuildTree();
+            }
+        }
+    }
+
+
+    private OctTree? CreateNode(BoundingBox bb, List<IIntersectable> objects)
+    {
+        if (objects.Count == 0)
+        {
+            return null;
+        }
+
+        return new OctTree(bb, objects)
+        {
+            // _parent = this,
+            _treeDepth = _treeDepth + 1
+        };
+    }
+
+    private void RecalculateBoundingBox()
+    {
+        //TODO: refactor to Pow 2
+        BoundingBox bb = _objects[0].BB;
+        for (int i = 1; i < _objects.Count; i++)
+        {
+            BoundingBox.Union(ref bb, _objects[i].BB);
+        }
+
+        BB = bb;
+    }
+
+    private BoundingBox[] CreateNewOctant()
+    {
+        Vector3 dimensions = BB.Max - BB.Min;
+        Vector3 half = dimensions / 2.0f;
+        Vector3 center = BB.Min + half;
+
+        BoundingBox[] octant = new BoundingBox[8];
+        for (int i = 0; i < 8; i++)
+        {
+            octant[i] = CreateNewOctant(i, center);
+        }
+
+        return octant;
+    }
+
+    private BoundingBox CreateNewOctant(int i, Vector3 center)
+    {
+        switch (i)
+        {
+            case 0: return new BoundingBox(BB.Min, center);
+            case 1:
+                return new BoundingBox(new Vector3(center.X, BB.Min.Y, BB.Min.Z),
+                    new Vector3(BB.Max.X, center.Y, center.Z));
+            case 2:
+                return new BoundingBox(new Vector3(center.X, BB.Min.Y, center.Z),
+                    new Vector3(BB.Max.X, center.Y, BB.Max.Z));
+            case 3:
+                return new BoundingBox(new Vector3(BB.Min.X, BB.Min.Y, center.Z),
+                    new Vector3(center.X, center.Y, BB.Max.Z));
+            case 4:
+                return new BoundingBox(new Vector3(BB.Min.X, center.Y, BB.Min.Z),
+                    new Vector3(center.X, BB.Max.Y, center.Z));
+            case 5:
+                return new BoundingBox(new Vector3(center.X, center.Y, BB.Min.Z),
+                    new Vector3(BB.Max.X, BB.Max.Y, center.Z));
+            case 6: return new BoundingBox(center, BB.Max);
+            case 7:
+                return new BoundingBox(new Vector3(BB.Min.X, center.Y, center.Z),
+                    new Vector3(center.X, BB.Max.Y, BB.Max.Z));
+            default:
+                throw new ArgumentException();
+        }
+    }
+
+
+    public void Transform(WorldTransform wt)
+    {
+        throw new NotImplementedException();
```

Comment:
We have to support this method, because OctTree can be a shape on a Scene, and we apply transformations to it (i.e. translation).
To make it simple I would suggest rebuilding the tree from scratch when this method is called